pkg/controller/certificates/keystore_test.go

```diff
@@ -0,0 +1,225 @@
+/*
+Copyright 2020 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificates
+
+import (
+	"bytes"
+	"testing"
+
+	jks "github.com/pavel-v-chernykh/keystore-go"
+	"software.sslmate.com/src/go-pkcs12"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+func mustGeneratePrivateKey(t *testing.T, encoding cmapi.KeyEncoding) []byte {
+	pk, err := pki.GenerateRSAPrivateKey(2048)
+	if err != nil {
+		t.Fatal(err)
+	}
+	pkBytes, err := pki.EncodePrivateKey(pk, encoding)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return pkBytes
+}
+
+func mustSelfSignCertificate(t *testing.T, pkBytes []byte) []byte {
+	if pkBytes == nil {
+		pkBytes = mustGeneratePrivateKey(t, cmapi.PKCS8)
+	}
+	pk, err := pki.DecodePrivateKeyBytes(pkBytes)
+	if err != nil {
+		t.Fatal(err)
+	}
+	x509Crt, err := pki.GenerateTemplate(&cmapi.Certificate{
+		Spec: cmapi.CertificateSpec{
+			DNSNames: []string{"example.com"},
+		},
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	certBytes, _, err := pki.SignCertificate(x509Crt, x509Crt, pk.Public(), pk)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return certBytes
+}
+
+func TestEncodeJKSKeystore(t *testing.T) {
+	tests := map[string]struct {
+		password               string
+		rawKey, certPEM, caPEM []byte
+		verify                 func(t *testing.T, out []byte, err error)
+	}{
+		"encode a JKS bundle for a PKCS1 key and certificate only": {
+			password: "password",
+			rawKey:   mustGeneratePrivateKey(t, cmapi.PKCS1),
+			certPEM:  mustSelfSignCertificate(t, nil),
+			verify: func(t *testing.T, out []byte, err error) {
+				if err != nil {
+					t.Errorf("expected no error but got: %v", err)
+					return
+				}
+				buf := bytes.NewBuffer(out)
+				ks, err := jks.Decode(buf, []byte("password"))
+				if err != nil {
+					t.Errorf("error decoding keystore: %v", err)
+					return
+				}
+				if ks["certificate"] == nil {
+					t.Errorf("no certificate data found in keystore")
+				}
+				if ks["ca"] != nil {
+					t.Errorf("unexpected ca data found in keystore")
+				}
+			},
+		},
+		"encode a JKS bundle for a PKCS8 key and certificate only": {
+			password: "password",
+			rawKey:   mustGeneratePrivateKey(t, cmapi.PKCS8),
+			certPEM:  mustSelfSignCertificate(t, nil),
+			verify: func(t *testing.T, out []byte, err error) {
```

Comment:
Can we somehow avoid having this validation func here so many time?

Reply:
The validation function is only the same in the first two cases, I feel like for the sake of keeping the 'verify' step written nearby the input parameters for each case, and given it's only here repeated twice, it's not so bad.

We can't just match `err`, as in the final P12 case we are verifying that the error returned when *decoding* the bundle is as expected, not the error from the function. Overall I'm just not sure it's worth the extra complexity for the sake of a handful of cases 🤷‍♂

Reply:
Okay, it feels repeated more than twice when reviewing it, but i agree that twice should be fine

Reply:
The two different `Test` functions decode using a different decoder in order to assert the returned bundled - one uses the p12 decoder and the other uses the JKS decoder - both look very similar 😅